### DISCUSSION TITLE
Async IValueConverter freezes application

### DISCUSSION THREAD

**Discussion by LSXAxeller** [AvaloniaUI/Avalonia#16901] [Q&A]
I'm trying to create an asynchronous converter in Avalonia, but I'm running into an issue where the program freezes when calling `.ConfigureAwait(false).GetAwaiter().GetResult()` or even just `.Result` on my async method.

Here's the code for my converter:

```csharp
using System;
// ... other using statements

public class ConditionMultiConverter : IValueConverter
{
    // ... other methods

    public object? Convert(object? value, Type targetType, object? parameter, CultureInfo culture)
    {
        // ... other cases

        return parameterString switch
        {
            // ... other cases
            "subtitleslanguages" => ConvertToSubtitlesLanguage(value?.ToString()).ConfigureAwait(false).GetAwaiter().GetResult(),
            _ => value
        };
    }

    // ... other methods

    private static async Task<List<string>> ConvertToSubtitlesLanguage(string? url)
    {
         if (string.IsNullOrEmpty(url))
            return [];

        if (!Uri.TryCreate(url, UriKind.Absolute, out var uri) || uri.IsFile ||
            !uri.AbsoluteUri.Contains("youtube.com")) return [];
        try
        {
            var youtube = new YoutubeClient();
            var trackManifest = await youtube.Videos.ClosedCaptions.GetManifestAsync(url);
            return trackManifest.Tracks.Select(track => track.Language.Code.ToUpperInvariant()).Distinct()
                .ToList();
        }
        catch
        {
            return [];
        }
    }
}
```

The `ConvertToSubtitlesLanguage` method fetches subtitles languages asynchronously using YoutubeExplode library.

**Here's how I'm using the converter in my XAML:**

```xml
<ComboBox ItemsSource="{Binding Link, Converter={StaticResource ConditionMultiConverter}, ConverterParameter='SubtitlesLanguages'}" />
```

I understand that blocking the UI thread with synchronous waits is a bad practice, but I'm unsure how to implement an asynchronous converter in Avalonia correctly.

**What I've Tried:**

* Using `.ConfigureAwait(false).GetAwaiter().GetResult()`: The program freezes.
* Using `.Result`: The program freezes.
* Implementing an Async Converter using `TaskCompletionNotifier` as suggested in this gist: [AsyncConverter.cs for WPF](https://gist.github.com/Platonenkov/4c2cb02dacc0fa0df1dcce6920a6c814): Still no luck.

Any help or insights would be greatly appreciated!

Thanks in advance!

**Comment by workgroupengineering** [comment]
See this https://docs.avaloniaui.net/docs/guides/data-binding/how-to-bind-to-a-task-result

**Reply by LSXAxeller** [reply]
This is working only for view models not converters, I can't use ^ with converters `<ComboBox ItemsSource="{Binding Path, Converter={StaticResource ConditionMultiConverter}, ConverterParameter='SubtitlesLanguages' ^}" />` gives `Unexpected token`

**Reply by IanRawley** [reply]
I think you need to put the ^ after Path and have the converter return the task directly, so {Binding Path^, ....

**Reply by LSXAxeller** [reply]
`<ComboBox ItemsSource="{Binding Path^, Converter={StaticResource ConditionMultiConverter}, ConverterParameter='SubtitlesLanguages'}" />`  

MemoryPageView.axaml(213, 63): [AVLN2000] Internal compiler error: Object reference not set to an instance of an object. (AvaloniaXamlIlBindingPathTransformer) Line 213, position 63.

**Comment by IanRawley** [comment]
You're blocking your UI thread with the GetAwaiter().GetResult() chain.  See https://devblogs.microsoft.com/dotnet/configureawait-faq/#i’m-using-getawaiter().getresult().-do-i-need-to-use-configureawait(false)11

ConfigureAwait(false) here doesn't do anything.  So you'll instead need to make your Convert method properly async, or find some other way to achieve the same results.

**Comment by YakudzaBY** [comment]
Atm using .Result is working

**Reply by stevemonaco** [reply]
It will work until it doesn't. I strongly suggest never trying any sort of async converters and instead do that work in the ViewModel. For things like images, rewriting the `Image` control is the right step.

**Reply by YakudzaBY** [reply]
We are using converters when we want to Avalonia track `INotify[Property|Collection]Change` in a long binding path chain automatically, leaving only last step(s) to shape result. Moving that repeatable logic to ViewModel is tedious, while shaping last non-trackable step of result might be async on it's own.
